### PR TITLE
Prototype wingsail ctrl

### PR DESCRIPTION
**Pull request by alberto-escobar**
### Description

- Evan and I have prototyped the wingsail controller algorithm in this notebook along with a plot to demonstrate how algorithm behaves in different wind conditions.

### Verification

- end of notebook has a diagram section that helps illustrate how computed angles will move the wing sail.
- Further verification will be needed

### Resources

- Evan's brain
- Alberto's brain

### CODE REVIEW

File: controller/prototype_wingsail_controller.ipynb
Comment (DFriend01) [inline controller/prototype_wingsail_controller.ipynb:1]: For `compute_reynolds_number`
- In the markdown cell for `compute_reynolds_number`, can you put the equation for Reynolds Number in latex?
- Can you explain where you got the numbers `CHORD_WIDTH_MAIN_SAIL` and `KINEMATIC_VISCOSITY`?

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: For all your functions, be sure to add docstrings. Also, avoid using `#` comments to document functions. They should be used to document inline code, but not function documentation.

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: For `compute_trim_tab_angle`:
- Rather than checking for exact equality to zero, check if the `wind_direction` variable is **close to** zero. Check out the function [`np.isclose`](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html). In reality, you will run into issues if you pass in floating point values and check for exact equality.
- Can you justify this function at a high level? For example, explain that the trim tab should point in the direction of the apparent wind after it stabilizes.
- Why do you divide by the absolute value of wind direction (this ties to justifying the function) ?
- You don't use the `boat_direction` argument. Where you planning on computing the apparent wind from the wind and boat directions?

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: Generally, throughout this notebook, I think you are using "wind" and "apparent wind" interchangeably. Make sure to go through the notebook and make that distinction because that is a small detail that is actually very important.

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: In the cell after `compute_trim_tab_angle`:
- Explain where the lookup table came from. It adds credibility to your work. Consider putting a preamble in the markdown cell above and maybe even put a table of the values (for good documentation AND to practice your markdown 😉)
- Why is the trim tab angle `nan`? There is a probable a bug somewhere.
- In your print statements, consider printing up to two decimal places.

I also noticed you have two cells with identical code. Consider making that logic into a function and then call that function for each test.

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: For the `Boat Diagram` section, my feedback is kinda FYI but I think it's a good exercise if you want to optimize your code and condense it. I recommend addressing this comment last in a separate commit.

First, you can utilize numpy arrays to take care of some of the more rudimentary operations (i.e. `translated_point` now becomes `point - center` when using numpy arrays).

Second, the rotation can be handled via matrix multiplication. If $R$ defines your rotation matrix parameterized by $\theta$, then $y = Rx$. You can even further optimize this function by passing in multiple points. Say if $x$ is a column vector, you can transform multiple points in a single operation:

$Y = XR = [x_1 x_2 \dots x_n]^T R$

Then, your transformed points are stored in the rows of $Y$.

Either way, please make sure to document the methodology in the markdown cell and/or the function documentation.

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: Add some plot labels and a legend in the visualization

**Review by DFriend01 — CHANGES_REQUESTED** [review]

File: controller/prototype_wingsail_controller.ipynb
Comment (alberto-escobar) [reply controller/prototype_wingsail_controller.ipynb:1]: > For `compute_trim_tab_angle`:
> 
> * Rather than checking for exact equality to zero, check if the `wind_direction` variable is **close to** zero. Check out the function [`np.isclose`](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html). In reality, you will run into issues if you pass in floating point values and check for exact equality.
> * Can you justify this function at a high level? For example, explain that the trim tab should point in the direction of the apparent wind after it stabilizes.
> * Why do you divide by the absolute value of wind direction (this ties to justifying the function) ?
> * You don't use the `boat_direction` argument. Where you planning on computing the apparent wind from the wind and boat directions?

we are currently doing this:
`    if apparent_wind_direction >= 0:
        trim_tab_angle = desired_alpha
    else:
        trim_tab_angle = -desired_alpha
`

is that acceptable?

Comment (DFriend01) [inline controller/prototype_wingsail_controller.ipynb:1]: `compute_trim_tab` can be equivalently written as `return math.copysign(desired_alpha, apparent_wind_angle)`. See [`math.copysign`](https://docs.python.org/3/library/math.html#math.copysign) for how it works.

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: This comment is FYI... you **do not** need to implement this feedback.

I noticed you print up to two decimal places with the following: `print("Reynolds number: ", round(reynolds, 2))`.

You can use [f-strings](https://www.geeksforgeeks.org/formatted-string-literals-f-strings-python/) to equivalently do this, and they are often much easier to use when forming complex strings:

```python
print(f"Reynolds number: {reynolds:.2f}")
```

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: When referring to variable names or functions in Markdown documentation, I recommend wrapping them in backticks for readability. For example, if I use backticks like this:

\`lookup_table\`

it renders like this:

`lookup_table`

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: Feedback on `rotatePointAroundPoint`:

**Angle of Rotation**
Are positive angles defined for clockwise rotation or counterclockwise rotation? Just examining your code, I suspect that it is defined for clockwise rotations. Be sure to specify this in your documentation.

Also, you can change your rotation matrix rather than multiplying the angle by -1 because it looks like you are using the rotation matrix for CCW rotations (hence why you multiply your angles by -1 to make it a CW rotation). Notice that:

$$
R_{\text{CW}}(\theta)
\=
R_{\text{CCW}}(-\theta) 
\=
\begin{bmatrix}
\cos(-\theta) & -\sin(-\theta) \\
\sin(-\theta) & \cos(-\theta)
\end{bmatrix}
\= 
\begin{bmatrix}
\cos(\theta) & \sin(\theta) \\
-\sin(\theta) & \cos(\theta)
\end{bmatrix}
$$

**Degrees to Radians Conversion**
I notice that you do this manually. There is actually a builtin function for this called `math.radians` (and `math.degrees` for the reverse conversion).

**Matrix Multiplication with numpy**
I notice that you use the following:

```python
rotated_point = np.dot(rotation_matrix, translated_point)
```

`np.dot` is typically used for dot products, but also happens to support matrix multiplication. However, it is recommended by the documentation to use the special `@` operator when doing matrix multiplication. It is defined by the numpy interface as "matrix multiplication". Don't confuse it with `*` because that is elementwise multiplication.

In other words, you can do this:

```python
rotated_point = rotation_matrix @ translated_point
```

but only if both `rotation_matrix` and `translated_point` are numpy arrays (which they are in your code).

Comment (DFriend01) [reply controller/prototype_wingsail_controller.ipynb:1]: This is a nitpick, but for `drawSail`, `showPlot`, and `drawBoat`, you say that they return `void`. There is no such thing as the `void` type in Python. I think what you mean to refer to is `None`, which is kinda like Python's version of "void" except for the fact that you can also assign variables as `None` in Python in addition to referring to it as a type.

**Review by DFriend01 — CHANGES_REQUESTED** [review]

**Comment by DFriend01** [comment]
Can you remove the commented out line that contains the old degrees-to-radians conversion?

**Review by DFriend01 — APPROVED** [review]
